### PR TITLE
Eliminate struct kvs_block which contained one member: an mblock id

### PR DESCRIPTION
**Pull request by alexttx**

### CODE REVIEW

File: lib/include/hse_ikvdb/blk_list.h [hunk 1]
```diff
@@ -8,23 +8,14 @@
...
- * @bk_blkid:  mblock id
- * @bk_handle: mblock handle
- */
-struct kvs_block {
-    uint64_t bk_blkid;
-};
-
 struct blk_list {
-    struct kvs_block *blks;
+    uint64_t *blks;
```
Comment (alexttx) [inline lib/include/hse_ikvdb/blk_list.h:12]: This summarizes the changes in this PR:  I replaced `struct kvs_block` with `uint64_t`.   In most cases there was an array of `struct kvs_block` which was converted to an array of `uint64_t`.  It is mostly a syntax change.  In some cases there is logical change I will point out.

File: lib/cn/blk_list.c [hunk 2]
```diff
@@ -92,26 +98,26 @@ blk_list_append(struct blk_list *blks, u64 blkid)
...
+        uint64_t *new_mbidv;
 
-        old_sz = sizeof(struct kvs_block) * blks->n_alloc;
-        grow_sz = sizeof(struct kvs_block) * BLK_LIST_PRE_ALLOC;
-        new = malloc(old_sz + grow_sz);
-        if (!new)
+        old_sz = sizeof(blks->blks[0]) * blks->n_alloc;
+        grow_sz = sizeof(blks->blks[0]) * BLK_LIST_PRE_ALLOC;
+        new_mbidv = malloc(old_sz + grow_sz);
+        if (!new_mbidv)
```
Comment (beckerg) [inline lib/cn/blk_list.c:106]: Opportunity to put the ev() in a useful place...

File: lib/cn/cn_tree.c [hunk 3]
```diff
@@ -1661,9 +1661,9 @@ cn_comp_commit(struct cn_compaction_work *w)
          */
         err = cndb_record_kvset_add(
                         w->cw_tree->cndb, tx, w->cw_tree->cnid,
-                        km.km_nodeid, &km, w->cw_kvsetidv[i], km.km_hblk.bk_blkid,
-                        w->cw_outv[i].kblks.n_blks, (uint64_t *)w->cw_outv[i].kblks.blks,
-                        w->cw_outv[i].vblks.n_blks, (uint64_t *)w->cw_outv[i].vblks.blks,
+                        km.km_nodeid, &km, w->cw_kvsetidv[i], km.km_hblk_id,
+                        w->cw_outv[i].kblks.n_blks, w->cw_outv[i].kblks.blks,
+                        w->cw_outv[i].vblks.n_blks, w->cw_outv[i].vblks.blks,
```
Comment (beckerg) [inline lib/cn/cn_tree.c:1666]: I thought you were going with mbid, vbid, kbid, hbid?

File: lib/cn/keep.c [hunk 4]
```diff
@@ -40,8 +40,8 @@ kvset_keep_vblocks(
 
     /* alloc both the vblks and the vbm; 1 free does both */
     sz = nv * sizeof(*vbm->vbm_blkv) + niv * sizeof(*vbm->vbm_map);
-    blks = calloc(1, sz);
-    if (!blks)
+    mem = calloc(1, sz);
+    if (!mem)
```
Comment (beckerg) [inline lib/cn/keep.c:44]: another ev()...

File: lib/cn/kvset.c [hunk 5]
```diff
@@ -371,60 +371,6 @@ kvset_kblk_init(
...
 }
 
-/**
- * blkid_list_to_vec()
- *
- * malloc a vector for blkids, and populate it from a blk_list.  Caller
- * must free the vector when finished.
- */
-static u64 *
-blkid_list_to_vec(struct blk_list *blkl, size_t bufc, u64 *bufv)
```
Comment (alexttx) [inline lib/cn/kvset.c:381]: This was a horrible function.  Glad to see it go.

File: lib/cn/keep.c [hunk 6]
```diff
@@ -55,13 +55,13 @@ kvset_keep_vblocks(
...
-        free(blks);
+        free(mem);
         return err;
     }
 
-    vbm->vbm_blkv = blks;
+    vbm->vbm_blkv = mem;
     vbm->vbm_blkc = 0;
-    vbm->vbm_map = (u32 *)(blks + nv);
+    vbm->vbm_map = (u32 *)(vbm->vbm_blkv + nv);
```
Comment (beckerg) [inline lib/cn/keep.c:64]: Is that cast still necessary?  Maybe change to uint32_t...

File: lib/cn/kvset.c [hunk 7]
```diff
@@ -371,60 +371,6 @@ kvset_kblk_init(
...
-}
-
-/* This is not a general purpose function!  The caller must ensure mapv is
- * large enough, and must also cleanup on failure.
- */
-static merr_t
-kvset_map_blklist(
-    struct mpool *            mp,
-    struct blk_list *         blks,
-    struct mpool_mcache_map **mapp)
```
Comment (alexttx) [inline lib/cn/kvset.c:405]: Another grotesque function gone.

File: lib/cn/kvset.c [hunk 8]
```diff
@@ -832,38 +775,27 @@ merr_t
...
-        u64 bufv[64];
-        u64 *idv;
-
-        idv = blkid_list_to_vec(&km->km_vblk_list, NELEM(bufv), bufv);
-        if (ev(!idv))
-            return merr(ENOMEM);
-
-        err = mbset_create(cn_tree_get_mp(tree), n_vblks, idv, sizeof(struct vblock_desc),
+    if (idc) {
+        err = mbset_create(cn_tree_get_mp(tree), idc, idv, sizeof(struct vblock_desc),
```
Comment (alexttx) [inline lib/cn/kvset.c:786]: Now we can stop creating temporary arrays of mblock IDs necessitated by mismatched APIs.

File: tests/mocks/repository/lib/mock_kvset.c [hunk 9]
```diff
@@ -181,32 +178,31 @@ mock_make_vblocks(struct kv_iterator **kvi, struct kvs_rparams *rp, int nv)
...
     km.km_vblk_list.n_alloc = nv;
-    km.km_vblk_list.blks = ds;
+    km.km_vblk_list.blks = vblk_ids;
 
     err = _make_common(kvi, 0, (struct mpool *)kvdata, rp, &km);
     if (err)
         free(kvdata);
 
-    free(ds);
+    free(vblk_ids);
```
Comment (beckerg) [inline tests/mocks/repository/lib/mock_kvset.c:205]: Seems like a lot of variability in the naming of various IDs variables...

**Review by beckerg — APPROVED** [review]

File: lib/cn/cn_tree.c (hunk 3 above)
Comment (alexttx) [reply lib/cn/cn_tree.c:1666]: I tried really hard to not rename the world.  There is one PR in this stack the renames `struct blk_list` members.

File: lib/cn/blk_list.c [hunk 10]
```diff
@@ -50,33 +52,37 @@ commit_mblocks(struct mpool *mp, struct blk_list *blks)
...
-    if (err) {
-        log_errx("Failed to delete mblock, blkid 0x%lx", err, blk->bk_blkid);
-        return err;
-    }
+    merr_t err;
 
-    blk->bk_blkid = 0;
+    INVARIANT(mp);
+
+    if (mbid) {
```
Comment (tristan957) [inline lib/cn/blk_list.c:61]: I would probably prefer an early return here. Doesn't mpool_mblock_delete() know how to handle a 0 mblock id?

File: lib/cn/blk_list.c [hunk 11]
```diff
@@ -50,33 +52,37 @@ commit_mblocks(struct mpool *mp, struct blk_list *blks)
...
+        err = mpool_mblock_delete(mp, mbid);
+        if (err) {
+            log_errx("Failed to delete mblock, blkid 0x%lx", err, mbid);
```
Comment (tristan957) [inline lib/cn/blk_list.c:64]: Maybe just `mblock (0x%lx)`

File: lib/include/hse_ikvdb/blk_list.h (hunk 1 above)
Comment (tristan957) [reply lib/include/hse_ikvdb/blk_list.h:12]: some form of this comment is worth putting in the final commit message :)

**Review by tristan957 — APPROVED** [review]

File: lib/cn/blk_list.c (hunk 10 above)
Comment (alexttx) [reply lib/cn/blk_list.c:61]: It does handle `mbid==0`

Comment (alexttx) [reply lib/cn/blk_list.c:61]: will update

**Pull request merged**